### PR TITLE
[Feature] Update multitask serving definition for YOLO.

### LINKED ISSUE #70: [Feature] Multihead serving with YOLO
# Prerequisites

Please answer the following question for yourself before submitting an issue.

- [x] I checked to make sure that this feature has not been requested already.

## 1. The entire URL of the file you are using

https://github.com/tensorflow/models/tree/master/official/vision/beta

## 2. Describe the feature you request

Add multihead inference definition for raw model, exported model and tflite.
Important to test the inference procedure or to match if phone implementation matches.

## 3. Additional context

Add any other context about the feature request here.

## 4. Are you willing to contribute it? (Yes or No)

Yes

### PR DESCRIPTION
**Pull request by rehohoho**
## Description
What feature/issue was addressed?
Multihead export module does not support YOLO head.

How was it resolved/added?
Add YOLO support in serving definition to export and run inference for multihead.
Refactor inference functions from yolo_ops into run_lib.

image

Any dependencies required for the change?

## Issue reference

Please reference the issue this PR will close: #70.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [ ] Other (Specify)

## Tests

Any reproducable method of testing to verify changes? \
List relevant details for the test configuration.

## Checklist
- [x] I have performed a self-review of my own code. Feel free to destroy my PR
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings.
- [ ] I have made corresponding changes to the documentation.
- [ ] I have added tests that prove my fix is effective or that my feature works.

### CODE REVIEW

**Comment by rehohoho** [comment]
Blocking #74 fix on testing branch.

File: official/vision/beta/serving/multitask.py [hunk 1]
```diff
@@ -1,15 +1,16 @@
...
 from official.vision.beta.modeling import factory_multitask
-from official.vision.beta.ops import preprocess_ops
+from official.vision.beta.ops import preprocess_ops, yolo_ops, box_ops
 from official.vision.beta.ops.colormaps import get_colormap
-from official.vision.beta.serving import export_base
+from official.vision.beta.serving import export_base, run_lib
+from official.vision.beta.projects.yolo.ops import box_ops as yolo_box_ops
 
 # RGB mean and stddev from ImageNet
 MEAN_RGB = (0.485 * 255, 0.456 * 255, 0.406 * 255)
```
Comment (anthonyfong100) [inline official/vision/beta/serving/multitask.py]: might be good to refactor all of this constants to a constants file

File: official/vision/beta/serving/multitask.py [hunk 2]
```diff
@@ -85,17 +87,139 @@ def serve(self, images: tf.Tensor) -> Mapping[str, tf.Tensor]:
...
+        processed_outputs[name + 'classes'] = classes
+        processed_outputs[name + 'scores'] = scores
+
+      else:
+        raise NotImplementedError('Task type %s is not implemented.' + \
+          'Try renaming the task routine.' %name)
 
     return processed_outputs
+  
+  def run(self,
```
Comment (anthonyfong100) [inline official/vision/beta/serving/multitask.py]: Small comment but it might be good to rename the function to show that it is running for an image directory

File: official/vision/beta/serving/multitask.py [hunk 3]
```diff
@@ -85,17 +87,139 @@ def serve(self, images: tf.Tensor) -> Mapping[str, tf.Tensor]:
...
+          image_path_glob: str,
+          output_dir: str,
+          preprocess_fn: Callable[[tf.Tensor], tf.Tensor],
```
Comment (anthonyfong100) [inline official/vision/beta/serving/multitask.py]: This is good love the type suggestions here

File: official/vision/beta/serving/multitask.py [hunk 4]
```diff
@@ -85,17 +87,139 @@ def serve(self, images: tf.Tensor) -> Mapping[str, tf.Tensor]:
...
+      class_names_path: `str`, path to txt file containing classes. Text file
+        should contain one class name per line.
+      save_logits_bin: `bool`, flag to save tensors and binary files
+    """
+    cmap = get_colormap(cmap_type='cityscapes').numpy()
+    dataset = run_lib.inference_dataset(image_path_glob=image_path_glob,
+                                        output_dir=output_dir,
+                                        preprocess_fn=preprocess_fn)
+
+    class_names_paths = class_names_path.split(',')
```
Comment (anthonyfong100) [inline official/vision/beta/serving/multitask.py]: This might potentially be brittle code here by loading the class names, try to refactor this out to another seperate function like `load_class_names`

File: official/vision/beta/serving/multitask.py [hunk 5]
```diff
@@ -85,17 +87,139 @@ def serve(self, images: tf.Tensor) -> Mapping[str, tf.Tensor]:
...
+                                    output_path=save_basename + '_boxes')
+        run_lib.write_tensor_as_bin(tensor=yolo_scores, 
+                                    output_path=save_basename + '_scores')
+        run_lib.write_tensor_as_bin(tensor=yolo_classes,
+                                    output_path=save_basename + '_classes')
+
+      image = tf.image.resize(image, self._input_image_size)
+      image = tf.cast(image, tf.uint8)
+
+      def tensor_to_numpy(tensor):
```
Comment (anthonyfong100) [inline official/vision/beta/serving/multitask.py]: This seems like a pretty common util function, might be good to move this to a seperate util file

File: official/vision/beta/serving/run_lib.py [hunk 6]
```diff
@@ -133,3 +138,97 @@ def get_export_module(experiment: str,
...
+  elif class_names_path.endswith('.json'):
+    with open(class_names_path) as f:
+      names = json.load(f)
+    if type(list(names.keys())[0]) == str and type(list(names.values())[0]) == int:
+      names = dict((v,k) for k,v in names.items())
+
+  else:
+    raise NotImplementedError('File type is not .txt or .json, path %s' %class_names_path)
+
+  assert type(list(names.keys())[0]) == int, 'Loaded dict %s has wrong key type %s' %(
```
Comment (anthonyfong100) [inline official/vision/beta/serving/run_lib.py]: This is a minor thing but try not to use assertions for non testing code, if python is run with the optimized flags this assertions might be dropped. Maybe we can just raise the error 

https://stackoverflow.com/questions/40182944/whats-the-difference-between-raise-try-and-assert

File: official/vision/beta/serving/run_lib.py [hunk 7]
```diff
@@ -133,3 +138,97 @@ def get_export_module(experiment: str,
...
+              num_bboxes: np.array,
+              class_names: Mapping[int, str], 
+              show_label: bool = True):
+  num_classes = len(class_names)
+  image_h, image_w, _ = image.shape
+  hsv_tuples = [(1.0 * x / num_classes, 1., 1.) for x in range(num_classes)]
+  colors = list(map(lambda x: colorsys.hsv_to_rgb(*x), hsv_tuples))
+  colors = list(map(lambda x: (int(x[0] * 255), int(x[1] * 255), int(x[2] * 255)), colors))
+
+  random.seed(0)
```
Comment (anthonyfong100) [inline official/vision/beta/serving/run_lib.py]: just curious why do we need to shuffle?

**Review by anthonyfong100 — COMMENTED** [review]
Left some comments mainly addressing the code quality aspect did not touch much about the core logic. Feel free to agree / disagree on points

**Comment by rehohoho** [comment]
I think all the comments are well raised, especially the one on the assertions. Did the relevant refactoring / documenting / changes.

File: official/vision/beta/serving/run_lib.py (hunk 7 above)
Comment (rehohoho) [reply official/vision/beta/serving/run_lib.py]: To fix a color palette according to seed. Forgot to make the seed an argument. Did so and documented in changes below.

**Review by anthonyfong100 — APPROVED** [review]
LGTM
